Report Requires diagnostics from dataflow analyzer

sbomer opened the pull request:
This moves the logic for generating diagnostics about RequiresUnreferencedCode,
RequiresAssemblyFiles, and RequiresDynamicCode attributes from the separate
Requires* analyzers into the DynamicallyAccessedMembersAnalyzer.

This includes logic for warning on access to attributed members, not the warnings
about mismatching attributes on base/override methods. The override validation
logic is still handled by the respective analyzers.

The DynamicallyAccessedMembersAnalyzer is now turned on by any of the settings
EnableTrimAnalyzer, EnableAotAnalyzer, or EnableSingleFileAnalyzer.

A new type RequiresAnalyzerContext is used to cache per-compilation state of
the analyzers, to avoid recomputing it for each trim analysis pattern that
might potentially produce a warning. The context only stores state for enabled
analyzers, based on the MSBuild property values.

There are a few minor differences in the warning behavior with this change:
- Implicit calls to annotated base ctors warn because they are visible in the CFG
- Unused local functions aren't discovered by the analysis, so don't warn
  (matching linker/AOT behavior)

sbomer committed:
Report Requires diagnostics from dataflow analyzer

This moves the logic for generating diagnostics about RequiresUnreferencedCode,
RequiresAssemblyFiles, and RequiresDynamicCode attributes from the separate
Requires* analyzers into the DynamicallyAccessedMembersAnalyzer.

This includes logic for warning on access to attributed members, not the warnings
about mismatching attributes on base/override methods. The override validation
logic is still handled by the respective analyzers.

The DynamicallyAccessedMembersAnalyzer is now turned on by any of the settings
EnableTrimAnalyzer, EnableAotAnalyzer, or EnableSingleFileAnalyzer.

A new type RequiresAnalyzerContext is used to cache per-compilation state of
the analyzers, to avoid recomputing it for each trim analysis pattern that
might potentially produce a warning. The context only stores state for enabled
analyzers, based on the MSBuild property values.

sbomer committed:
Merge remote-tracking branch 'origin/main' into unifyAnalyzers

sbomer committed:
Merge remote-tracking branch 'origin/main' into unifyAnalyzers

vitek-karas commented on src/tools/illink/src/ILLink.RoslynAnalyzer/RequiresAnalyzerBase.cs:
```diff
@@ -54,63 +55,11 @@ public override void Initialize (AnalysisContext context)
-					}
-				}, OperationKind.ObjectCreation);
-
 				context.RegisterOperationAction (operationContext => {
 					var fieldReference = (IFieldReferenceOperation) operationContext.Operation;
 					CheckCalledMember (operationContext, fieldReference.Field, incompatibleMembers);
```
This seems to be the last peice which needs CheckCalledMember (and all the infra that brings in). Would it make sense to move this to the DAM analyzer as well? I wonder how this plays with the rest of it. Don't we see field references in code in assignments and such anyway?

vitek-karas commented on src/tools/illink/src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs:
```diff
@@ -70,11 +70,16 @@ public sealed class RequiresUnreferencedCodeAnalyzer : RequiresAnalyzerBase
+			out Diagnostic? incompatibleMembersDiagnostic)
 		{
+			incompatibleMembersDiagnostic = null;
 			// Some RUC-annotated APIs are intrinsically handled by the trimmer
 			if (member is IMethodSymbol method && Intrinsics.GetIntrinsicIdForMethod (new MethodProxy (method)) != IntrinsicId.None) {
 				return true;
```
This is unfortunate - that said we have a similar hack in the trimmer. I guess improvement in the future - how to handle RUC and similar correctly with the intrinsics handling.

vitek-karas replied:
That is more the point of the comment here: https://github.com/dotnet/runtime/pull/92724#discussion_r1355251455

This one is about the fact that we have a "External" way of suppressing RUC warnings on intrinsics. In here, we check if a method is intrinsic and then trust the other analyzer it will do its job instead. In linker we use this giant enum which describes which pattern the warning came from, and then based on that we suppress the "general" RUC processing and rely on the method body scanner instead.

NativeAOT is the only implementation where this weird approach is not used - there it's all part of the main "handle call" functionality (still not ideal where it is, but it's much better), and we don't need a way to determine if something will be handled by the intrinsics or other mechanics to decide if we should apply RUC logic or not.

This is definitely not something to fix in this PR though.

vitek-karas reviewed (approved):
Looks good - the only problem I see is the potentially problematic suppression of warnings due to RUC/RDC/RAF in presense of intrinsic handling of AOT specific behaviors. But I don't think we have such a case yet, so it doesn't block functionality... yet.

sbomer committed:
Add test for annotated value assigned to event

sbomer committed:
Use DiagnosticContext for consistency

sbomer committed:
Move field access handling into dataflow analyzer

sbomer committed:
Remove unnecessary merge logic for reflection access pattern

vitek-karas commented on src/tools/illink/src/ILLink.RoslynAnalyzer/TrimAnalysis/TrimAnalysisFieldAccessPattern.cs:
```diff
@@ -0,0 +1,50 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices.ComTypes;
+using ILLink.Shared;
+using ILLink.Shared.DataFlow;
+using ILLink.Shared.TrimAnalysis;
+using ILLink.Shared.TypeSystemProxy;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Diagnostics;
+using MultiValue = ILLink.Shared.DataFlow.ValueSet<ILLink.Shared.DataFlow.SingleValue>;
```
Nit - these seem excessive - could be probably trimmed down.

vitek-karas commented on src/tools/illink/src/ILLink.RoslynAnalyzer/TrimAnalysis/TrimAnalysisReflectionAccessPattern.cs:
```diff
@@ -34,23 +35,23 @@ namespace ILLink.RoslynAnalyzer.TrimAnalysis
-
-			return new TrimAnalysisReflectionAccessPattern (
-				ReferencedMethod,
-				lattice.Meet (Instance, other.Instance),
```
The Instance property doesn't seem to be used - we should remove it. It was the sole reason for having a Merge... without it there's really no reason to have Merge.

sbomer committed:
PR feedback

- Clean up a few unused usings
- Remove unused Instance field

sbomer committed:
Merge remote-tracking branch 'origin/main' into unifyAnalyzers

sbomer committed:
Fix ExpectedWarnings for ILLink

sbomer committed:
Fix assert for ref property as out param

And related cases involving implicit indexer references.

sbomer committed:
Add test for deconstruction assignment to ref property

sbomer committed:
Fix delegate creation over invocation

sbomer committed:
Fix ExpectedWarnings for illink

sbomer committed:
Silence dataflow warnings if EnableTrimAnalyer isn't set

Extra warnings were being produced on some projects where
the single-file analyzer was enabled, causing the DAM analyzer
to light up. The dataflow warnings were missing a check of
EnableTrimAnalzer to prevent these warnings.

This also lifts the checking of MSBuild property values out into
the compilation start action, so it only needs to be done per
compilation. The context type has been renamed to reflect that
it holds state relevant to all of the analyses handled by the
dataflow logic now.

sbomer committed:
Add test coverage for EnableTrimAnalyzer checks

And add missing checks in DAM analyzer

vitek-karas commented:
Looks good - thanks!
